src/native/clr/host/gc-bridge.cc

```diff
@@ -0,0 +1,315 @@
+}
+
+void GCBridge::trigger_java_gc () noexcept
+{
+	env->CallVoidMethod (Runtime_instance, Runtime_gc);
+	if (env->ExceptionCheck ()) [[unlikely]] {
+		env->ExceptionDescribe ();
+		env->ExceptionClear ();
+		log_error (LOG_DEFAULT, "Java GC failed");
+	}
+}
+
+void GCBridge::prepare_for_java_collection (MarkCrossReferencesArgs* cross_refs) noexcept
+{
+	// Some SCCs might have no IGCUserPeers associated with them, so we must create one
+	std::unordered_map<StronglyConnectedComponent*, jobject> temporary_peers;
+
+	// Before looking at xrefs, scan the SCCs. During collection, an SCC has to behave like a
+	// single object. If the number of objects in the SCC is anything other than 1, the SCC
+	// must be doctored to mimic that one-object nature.
+	for (size_t i = 0; i < cross_refs->ComponentCount; i++) {
+		StronglyConnectedComponent *scc = &cross_refs->Components [i];
+
+		// Count > 1 case: The SCC contains many objects which must be collected as one.
+		// Solution: Make all objects within the SCC directly or indirectly reference each other
+		if (scc->Count > 1) {
+			add_references (scc);
+		} else if (scc->Count == 0) {
+			temporary_peers [scc] = env->NewObject (GCUserPeer_class, GCUserPeer_ctor);
+		}
+	}
+
+	// Add the cross scc refs
+	for (size_t i = 0; i < cross_refs->CrossReferenceCount; i++) {
+		ComponentCrossReference *xref = &cross_refs->CrossReferences [i];
+
+		StronglyConnectedComponent *source = &cross_refs->Components [xref->SourceGroupIndex];
+		jobject from = source->Count > 0 ? source->Contexts [0]->control_block->handle : temporary_peers [source];
+		
+		StronglyConnectedComponent *destination = &cross_refs->Components [xref->DestinationGroupIndex];
+		jobject to = destination->Count > 0 ? destination->Contexts [0]->control_block->handle : temporary_peers [destination];
+
+		add_reference (from, to);
+	}
+
+	// With cross references processed, the temporary peer list can be released
+	for (const auto& [scc, temporary_peer] : temporary_peers) {
+		env->DeleteLocalRef (temporary_peer);
+	}
+
+	// Switch global to weak references
+	for (size_t i = 0; i < cross_refs->ComponentCount; i++) {
+		StronglyConnectedComponent *scc = &cross_refs->Components [i];
+		for (size_t j = 0; j < scc->Count; j++) {
+			take_weak_global_ref (scc->Contexts [j]);
+		}
+	}
+}
+
+void GCBridge::add_references (StronglyConnectedComponent *scc) noexcept
+{
+	abort_if_invalid_pointer_argument (scc, "scc");
+	abort_unless (scc->Count > 1, "SCC must have at least two items to add inner references");
+
+	JniObjectReferenceControlBlock *prev = scc->Contexts [scc->Count - 1]->control_block;
+
+	for (size_t j = 1; j < scc->Count; j++) {
+		JniObjectReferenceControlBlock *current = scc->Contexts [j]->control_block;
+		if (add_reference (prev->handle, current->handle)) {
+			prev->refs_added++;
+		}
+
+		prev = current;
+	}
+}
+
+bool GCBridge::add_reference (jobject from, jobject to) noexcept
+{
+	jclass java_class = env->GetObjectClass (from);
```

Comment:
We either need to validate `from` and `to` here or handle potential Java exceptions/error conditions below. I'd prefer verifying the pointers here.